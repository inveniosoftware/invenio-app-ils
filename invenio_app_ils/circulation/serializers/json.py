# -*- coding: utf-8 -*-
#
# Copyright (C) 2018-2020 CERN.
#
# invenio-app-ils is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Loan JSON serializers."""

from invenio_records_rest.serializers.json import JSONSerializer

from invenio_app_ils.literature.serializers.custom_fields import \
    field_cover_metadata

from .custom_fields import (field_is_overdue, field_pickup_location,
                            field_transaction_location, field_transaction_user)


class LoanJSONSerializer(JSONSerializer):
    """Serialize loan."""

    def transform_record(self, pid, record, links_factory=None, **kwargs):
        """Transform record into an intermediate representation."""
        loan = super().transform_record(
            pid, record, links_factory=links_factory, **kwargs
        )
        field_is_overdue(loan["metadata"])
        field_pickup_location(loan["metadata"])
        field_transaction_location(loan["metadata"])
        field_transaction_user(loan["metadata"])
        field_cover_metadata(loan["metadata"].get("document", {}))
        return loan

    def transform_search_hit(
        self, pid, record_hit, links_factory=None, **kwargs
    ):
        """Transform search result hit into an intermediate representation."""
        hit = super().transform_search_hit(
            pid, record_hit, links_factory=links_factory, **kwargs
        )
        field_is_overdue(hit["metadata"])
        field_cover_metadata(hit["metadata"].get("document", {}))
        return hit
