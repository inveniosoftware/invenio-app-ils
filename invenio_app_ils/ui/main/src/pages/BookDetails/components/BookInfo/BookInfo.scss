$padding-40: 40px;
$margin-30: 30px;
$color-author: #b1b2ac;

.book-details-info {
  display: flex;
  flex-direction: column;

  padding: 0 $padding-40;
}

.book-details-author {
  color: $color-author;
}

.book-details-abstract {
  margin-top: $margin-30;
}
