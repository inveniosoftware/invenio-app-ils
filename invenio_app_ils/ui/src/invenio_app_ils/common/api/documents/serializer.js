import isEmpty from 'lodash/isEmpty';
import { fromISO } from '../date';

function serializeSeries(series) {
  return {
    serial: series.filter(s => s.mode_of_issuance === 'SERIAL'),
    multipart: series.filter(s => s.mode_of_issuance === 'MULTIPART_MONOGRAPH'),
  };
}

function serializeResponse(hit) {
  let result = {};
  if (!isEmpty(hit)) {
    result['id'] = hit.id;
    result['created'] = fromISO(hit.created);
    result['updated'] = fromISO(hit.updated);
    if (hit.links) {
      result['links'] = hit.links;
    }
    if (!isEmpty(hit.metadata)) {
      result['metadata'] = hit.metadata;
      if (!isEmpty(hit.metadata.circulation.next_available_date)) {
        result['metadata']['circulation']['next_available_date'] = new Date(
          result['metadata']['circulation']['next_available_date']
        ).toDateString();
      }
      if (!isEmpty(hit.metadata.series)) {
        // Split the series based on mode of issuance - requested by the library
        result['metadata']['series'] = serializeSeries(hit.metadata.series);
      }
      result['pid'] = hit.metadata.pid;
    }
  }
  return result;
}

export const serializer = {
  fromJSON: serializeResponse,
};
