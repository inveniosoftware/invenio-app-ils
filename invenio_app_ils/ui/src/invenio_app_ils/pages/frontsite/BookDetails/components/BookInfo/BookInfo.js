import React, { Component } from 'react';
import { Header, List } from 'semantic-ui-react';

export default class BookInfo extends React.Component {
  render() {
    return (
      <div className="book-info">
        <Header as="h2">Title: {this.props.data.title}</Header>
        <List>
          {this.props.data.authors.map(author => (
            <List.Item as="h4">Author: {author}</List.Item>
          ))}
        </List>
        <List>
          {this.props.data.publishers.map(publisher => (
            <List.Item as="h5">Publisher: {publisher}</List.Item>
          ))}
        </List>
      </div>
    );
  }
}
