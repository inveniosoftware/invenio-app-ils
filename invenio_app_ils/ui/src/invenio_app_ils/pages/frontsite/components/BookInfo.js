import React, { Component } from 'react';
import { Header, List } from 'semantic-ui-react';
import PropTypes from 'prop-types';

export class BookInfo extends Component {
  render() {
    const { documentMetadata } = this.props;
    return (
      <div className="document-info">
        <Header as="h2">{documentMetadata.title.title}</Header>
        <List>
          {documentMetadata.authors.map((author, index) => (
            <List.Item as="h4" key={`Key${index}`}>
              Author: {author.full_name}
            </List.Item>
          ))}
        </List>
      </div>
    );
  }
}

BookInfo.propTypes = {
  documentMetadata: PropTypes.object.isRequired,
};
