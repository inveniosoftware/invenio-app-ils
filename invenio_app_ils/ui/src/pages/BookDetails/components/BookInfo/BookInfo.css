.book-details-info {
  display: flex;
  flex-direction: column;
  padding: 0 40px; }

.book-details-author {
  color: #b1b2ac; }

.book-details-abstract {
  margin-top: 30px; }
