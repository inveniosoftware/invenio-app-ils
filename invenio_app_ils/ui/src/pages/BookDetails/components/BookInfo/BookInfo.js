import React, { Component } from 'react';
import BookTabs from '../BookTabs';

import './BookInfo.css';
import { Loader } from 'semantic-ui-react';

class BookInfo extends Component {
  renderBookAuthors(data) {
    if (data.authors) {
      return data.authors.map(author => author.full_name).join();
    } else if (data.corporate_authors) {
      return data.corporate_authors.join();
    } else {
      return 'No authors available';
    }
  }

  render() {
    let { data, isLoading } = this.props;
    if (isLoading) return <Loader active inline="centered" />;
    return (
      <div className="book-details-info">
        <div className="book-details-title">
          <h1>{data.titles[0].title}</h1>
        </div>
        <div className="book-details-author">
          by {this.renderBookAuthors(data)}
        </div>
        <div className="book-details-abstract">{data.abstracts[0]}</div>
        <div className="book-details-other">Aditional Information</div>
        <BookTabs />
      </div>
    );
  }
}

export default BookInfo;
