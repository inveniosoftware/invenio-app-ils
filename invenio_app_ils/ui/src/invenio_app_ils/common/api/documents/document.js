import { http } from '../base';
import { serializer } from './serializer';

const documentURL = '/documents/';

const get = documentPid => {
  return http.get(`${documentURL}${documentPid}`).then(response => {
    response.data = serializer.fromJSON(response.data);
    return response;
  });
};

class QueryBuilder {
  constructor() {
    this.overbookedQuery = [];
    this.availableItemsQuery = [];
    this.withPendingLoansQuery = [];
  }

  overbooked() {
    this.overbookedQuery.push(`circulation.overbooked:true`);
    return this;
  }

  withAvailableItems() {
    this.availableItemsQuery.push('circulation.loanable_items:>0');
    return this;
  }

  withPendingLoans() {
    this.withPendingLoansQuery.push('circulation.pending_loans:>0');
    return this;
  }

  qs() {
    return this.overbookedQuery
      .concat(this.availableItemsQuery, this.withPendingLoansQuery)
      .join(' AND ');
  }
}

const queryBuilder = () => {
  return new QueryBuilder();
};

const list = query => {
  return http.get(`${documentURL}?q=${query}`).then(response => {
    const totalHits = response.data.hits.total;
    response.data = response.data.hits.hits.map(hit =>
      serializer.fromJSON(hit)
    );
    response.data.totalHits = totalHits;
    return response;
  });
};

const count = query => {
  return http.get(`${documentURL}?q=${query}`).then(response => {
    response.data = response.data.hits.total;
    return response;
  });
};

export const document = {
  get: get,
  list: list,
  count: count,
  query: queryBuilder,
  serializer: serializer,
  url: documentURL,
};
