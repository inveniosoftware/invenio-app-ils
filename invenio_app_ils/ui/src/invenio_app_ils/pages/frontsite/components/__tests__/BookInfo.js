import React from 'react';
import renderer from 'react-test-renderer';
import { BookInfo } from '../BookInfo';
import * as testData from '../../../../../../../../tests/data/documents.json';

it('should render correctly when there are no publishers', () => {
  const documentMetadata = {
    ...testData[0],
  };
});

it('should render correctly when there are no authors', () => {
  const documentMetadata = {
    ...testData[0],
  };
});

it('should render correctly', () => {
  const documentMetadata = {
    ...testData[9],
  };
  const tree = renderer
    .create(<BookInfo documentMetadata={documentMetadata} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
