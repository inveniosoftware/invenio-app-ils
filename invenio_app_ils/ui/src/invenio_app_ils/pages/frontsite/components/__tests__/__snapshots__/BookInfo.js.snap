// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="document-info"
>
  <h2
    className="ui header"
  >
    The Swiss years : writings, 1900-1909
  </h2>
  <div
    className="ui list"
    role="list"
  >
    <h4
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      Author: 
      Einstein, Albert
    </h4>
    <h4
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      Author: 
      Stachel, John J et al.
    </h4>
  </div>
</div>
`;
